### PR TITLE
[ipv4] Unit Tests for IPv4 Layer

### PR DESCRIPTION
**Pull request by ppenna**
Description
========

This Pull Request addresses the following points:

- [x] Rename `fd` to `qd` in high-level calls to match Demikernel's terminology (8b929693d97cb7edd8aad37f72e409f877ca6697)
- [x] Add tracing statements to `pushto` and `pushto2` to ease debug in the UDP stack (ac0e8961ccc1c5b5d88a3216f2b5ed75bb40b892)
- [x] Use slicing to resize IPv4 datagrams after parsing procedure, and thus make the code cleaner to read (9580cfeeaf7056d306d99e321499d130ff721fef)
- [x] Introduce unit tests for happy, invalid, and unsupported paths in the IPv4 layer (a969bcf11b1c4211193979493e0b2cbda02e8ab6).

Related Issues
=========

- #115

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -213,23 +213,23 @@ impl<RT: Runtime> Catnip<RT> {
...
         timer!("catnip::accept");
-        trace!("accept(): {:?}", fd);
-        let r = match self.file_table.get(fd) {
+        trace!("accept(): {:?}", qd);
+        let r = match self.file_table.get(qd) {
             Some(qtype) => match QType::try_from(qtype) {
                 Ok(QType::TcpSocket) => {
-                    let newfd = self.file_table.alloc(QType::TcpSocket.into());
-                    Ok(FutureOperation::from(self.ipv4.tcp.do_accept(fd, newfd)))
+                    let newqd = self.file_table.alloc(QType::TcpSocket.into());
```
Comment (BrianZill) [inline src/lib.rs:231]: If we're going to try to standardize on rust_style variable names, then shouldn't this be "new_qd"?

File: src/protocols/ipv4/datagram.rs [hunk 2]
```diff
@@ -239,7 +234,19 @@ impl Ipv4Header {
             dst_addr,
         };
 
-        Ok((header, buf))
+        // Truncate payload.
+        Ok((
+            header,
+            &buf[(IPV4_HEADER_MIN_SIZE as usize)..(total_length as usize)],
```
Comment (BrianZill) [inline src/protocols/ipv4/datagram.rs:240]: There is an existing bug we're glossing over here.  We should be taking the IPv4 header length from the incoming packet, verifying that it is reasonable, and trimming that amount off the front to get to the payload.

File: src/protocols/ipv4/tests.rs [hunk 3]
```diff
@@ -1,97 +1,655 @@
...
 //==============================================================================
-// IPv4 Ping
+// Helper Functions
 //==============================================================================
 
+// Builds an IPv4 header.
+fn build_ipv4_header(
+    buf: &mut [u8],
+    version: u8,
+    ihl: u8,
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:22]: IHL (Internet Header Length?) isn't a common acronym in networking circles.  Probably should spell this out as header_length.

File: src/protocols/icmpv4/tests.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    .unwrap();
+
+    now += Duration::from_secs(1);
+    alice.rt().advance_clock(now);
+    bob.rt().advance_clock(now);
+
+    // Bob receives ping request from Alice.
+    bob.receive(alice.rt().pop_frame()).unwrap();
+
+    // Bob replies Alice.
```
Comment (BrianZill) [inline src/protocols/icmpv4/tests.rs:37]: Missing a "to" in "Bob replies to Alice."

File: src/protocols/icmpv4/tests.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    bob.rt().poll_scheduler();
+
+    now += Duration::from_secs(1);
+    alice.rt().advance_clock(now);
+    bob.rt().advance_clock(now);
+
+    // Alice receives reply from Bob
```
Comment (BrianZill) [inline src/protocols/icmpv4/tests.rs:44]: Missing the period at the end of "Alice receives reply from Bob."

File: src/protocols/ipv4/tests.rs [hunk 6]
```diff
@@ -1,97 +1,655 @@
...
+    fragment_flags: u8,
+    fragment_offset: u16,
+    ttl: u8,
+    protocol: u8,
+    src_addr: &[u8],
+    dest_addr: &[u8],
+    mut checksum: Option<u16>,
+) {
+    // Version + IHL
+    buf[0] = ((version & 0xf) << 4) | (ihl & 0xf);
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:36]: I'm of two minds on this one.  Yes, it is clearer to AND version with 0xf to show that you only care about the lower four bits before shifting it left 4.  But the shift left 4 with an 8 bit field will lose the 4 upper bits without the extra op.  So you don't need the AND, it is superfluous.  [Or does Rust actually check that you're potentially throwing meaningful bits away?  And the compiler generates a warning?  That would be pretty interesting, but I would be surprised if it is doing that.]  Since this is test code, I think it is okay to leave it in.  But for mainline code, I would be concerned that the compiler might not realize that it could optimize the AND out, and leave us executing some unnecessary instructions.

File: src/protocols/ipv4/tests.rs [hunk 7]
```diff
@@ -1,97 +1,655 @@
...
+
+    // DSCP + ECN
+    buf[1] = ((dscp & 0x3f) << 2) | (ecn & 0x3);
+
+    // Total Length
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:41]: Some of these comment lines end with a period, others don't.  I happen to think that (almost) all comments should be written in complete sentences that end with a period, but at the very least be consistent.

File: src/protocols/ipv4/tests.rs [hunk 8]
```diff
@@ -1,97 +1,655 @@
...
+        4,
+        5,
+        0,
+        0,
+        DATAGRAM_SIZE as u16,
+        0,
+        0x2,
+        0,
+        1,
+        IpProtocol::UDP as u8,
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:96]: This isn't related to your code per se, but I was surprised to see the "as u8" here.  We should define the IpProtocol enum as always being of type u8:

#[repr(u8)]
enum IpProtocol {
    ICMP = 1,
    TCP = 6,
    UDP =17,
...
}

The official protocol number registry is at https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml

File: src/protocols/ipv4/tests.rs [hunk 9]
```diff
@@ -1,97 +1,655 @@
 // Copyright (c) Microsoft Corporation.
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:1]: A meta comment on this file: A number of these tests are testing that we recognize "unsupported" values of certain fields in the packets we receive as such.  For most of these tests, this is okay, as most of these are checking for unsupported things (like say fragmented packets) that the spec requires us to support.  But there are also optional things (like say ECN support) that are optional in IP.  Our parser should parse IP headers with (for example) the "ECN-capable transport" bit set without complaint.  This is because that is an optional offering that we should simply ignore if we don't support ECN.

File: src/protocols/icmpv4/tests.rs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        _ => Err(()),
+    }
+    .unwrap();
+
+    now += Duration::from_secs(1);
+    alice.rt().advance_clock(now);
+    bob.rt().advance_clock(now);
+
+    // Bob receives ping request from Alice.
+    bob.receive(alice.rt().pop_frame()).unwrap();
```
Comment (BrianZill) [inline src/protocols/icmpv4/tests.rs:35]: Okay, you're probably not going to like this one, but I don't think we should use unwrap in test code, at least not in places like this.  And that's because it will cause a panic, rather than a "test failed" notification if it fails.  I recently had trouble trying to track down where a test was failing because it had an .unwrap instead of something that doesn't panic on failure and properly informs the test infrastructure that the test failed at this point.

Basically, test code shouldn't assume that something doesn't fail.  It should always explicitly check that things are working as we expected.  So we shouldn't use anything in test code that could panic.

File: src/lib.rs (hunk 1 above)
Comment (ppenna) [reply src/lib.rs:231]: Sure.

Comment (ppenna) [reply src/lib.rs:231]: Sure.

File: src/protocols/ipv4/datagram.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/ipv4/datagram.rs:240]: I agree. However, if I fix this we end up with failing tests for the TCP stack. Since this is a side bug, I'd suggest to keep track of it in a separate issue and PR. What do you think?

File: src/protocols/ipv4/tests.rs (hunk 3 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:22]: The RFC 791 refers to the internet header length as IHL (https://datatracker.ietf.org/doc/html/rfc791#section-3.1).

File: src/protocols/ipv4/tests.rs (hunk 6 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:36]: In the mainline code we don't have this operation, but I intentionally inserted this in test code to sanity check any optization that the compiler may be performing or not. I'd keep this in the testing code if you agree.

File: src/protocols/ipv4/tests.rs (hunk 9 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:1]: Agree. May we track this in another issue+PR?

File: src/protocols/ipv4/tests.rs (hunk 8 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:96]: The enum struct is defined to represented as a u8 type by the compiler, however, rust does not make any implicit data conversions, even if they are of same underlying representation. That is why we need the type conversion here.

File: src/protocols/icmpv4/tests.rs (hunk 4 above)
Comment (ppenna) [reply src/protocols/icmpv4/tests.rs:37]: sure

File: src/protocols/icmpv4/tests.rs (hunk 5 above)
Comment (ppenna) [reply src/protocols/icmpv4/tests.rs:44]: Sure

File: src/protocols/icmpv4/tests.rs (hunk 10 above)
Comment (ppenna) [reply src/protocols/icmpv4/tests.rs:35]: I'm unsure what you mean. If we assert the result that also causes the test to panic. Either way, cargo keeps running other tests and signals the failling one accordingly:

```
ppenna@mybox:~/demikernel/catnip$ make test TEST=ipv4_ping
/home/ppenna/.cargo/bin/cargo test --release  ipv4_ping -- --nocapture
   Compiling catnip v0.8.0 (/home/ppenna/demikernel/catnip)
    Finished release [optimized] target(s) in 5.41s
     Running unittests (target/release/deps/catnip-91fd64f5207022fc)

running 2 tests
[flexi_logger][ERRCODE::Time] flexi_logger has to work with UTC rather than with local time, caused by IndeterminateOffset
    See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#time
thread 'protocols::icmpv4::tests::ipv4_ping' panicked at 'assertion failed: false', src/protocols/icmpv4/tests.rs:29:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test protocols::icmpv4::tests::ipv4_ping ... FAILED
test protocols::icmpv4::tests::ipv4_ping_loop ... ok
```

File: src/protocols/ipv4/tests.rs (hunk 3 above)
Comment (BrianZill) [reply src/protocols/ipv4/tests.rs:22]: Okay.  Didn't remember that.

File: src/protocols/ipv4/tests.rs (hunk 8 above)
Comment (BrianZill) [reply src/protocols/ipv4/tests.rs:96]: Ah, so Rust thinks of this as a type conversion.  Interesting, and also too bad.  I think of enums as collections/wrappers that contain another type (like arrays, or even structs), so when specifying a single entry out of that collection, that entry should be of its own type, and not of some enum type.  But I guess if Rust requires it, we'll have to leave it.

File: src/protocols/ipv4/tests.rs (hunk 9 above)
Comment (BrianZill) [reply src/protocols/ipv4/tests.rs:1]: Yeah, we don't have to address this here and now.

Comment (ppenna) [reply src/protocols/ipv4/tests.rs:1]: Cool. I've created an issue for this: https://github.com/demikernel/catnip/issues/117

File: src/protocols/ipv4/tests.rs (hunk 8 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:96]: Yeah, there is no way to not cast it. Unlink C, rust is strongly typed (there is no implicit type coercion).

File: src/protocols/icmpv4/tests.rs (hunk 10 above)
Comment (ppenna) [reply src/protocols/icmpv4/tests.rs:35]: I see. So, what about we move to this strategy on-the-fly?

File: src/protocols/ipv4/datagram.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/ipv4/datagram.rs]: I've opened an issue for this: https://github.com/demikernel/catnip/issues/118

File: src/protocols/ipv4/tests.rs (hunk 8 above)
Comment (BrianZill) [reply src/protocols/ipv4/tests.rs:96]: Well, my objection with Rust is that it considers this a type conversion.  If we had a static array of u8's called IpProtocol, and referred to the entries via defined constants such as UDP = x, then Rust would consider IpProtocol[UDP] to be a u8 (I hope at least).  What type does it consider our enum entry IpProtocol::UDP to be I wonder?  Okay, I just checked this on the Rust playground, and it considers IpProtocol to be its own type.  I can see if one wants to ensure that one only assigns a valid IpProtocol entry to this field that this might be desired behavior.

So now I have a different objection: build_ipv4_header should take a IpProtocol here, not a u8.

File: src/protocols/ipv4/tests.rs [hunk 11]
```diff
@@ -1,97 +1,655 @@
...
+    version: u8,
+    ihl: u8,
+    dscp: u8,
+    ecn: u8,
+    total_length: u16,
+    fragment_id: u16,
+    fragment_flags: u8,
+    fragment_offset: u16,
+    ttl: u8,
+    protocol: u8,
```
Comment (BrianZill) [inline src/protocols/ipv4/tests.rs:30]: Shouldn't protocol be a IpProtocol, and not a u8?  Similarly, some of these other fields (like src_addr and dest_addr) are also more complex types that we're referring to here by their underlying representations.

Comment (ppenna) [reply src/protocols/ipv4/tests.rs:30]: If we do so, we cannot test bad values for this, because the `try_from` operation would complain. It is intentionally a `u8` to enable us to create malformed packages.

File: src/protocols/ipv4/tests.rs (hunk 8 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:96]: The function I wrote in the test function to intentionally test for badly received calls. There is no way to do it otherwise.

File: src/protocols/ipv4/tests.rs (hunk 11 above)
Comment (ppenna) [reply src/protocols/ipv4/tests.rs:30]: The rationale is simple: construct an IPv4 header using raw data (and without using the provided constructors in the IPv4 module). This way, we can be sure that the function for parsing is doing its job, not matter what it gets as input buffer.

File: src/protocols/icmpv4/tests.rs (hunk 10 above)
Comment (BrianZill) [reply src/protocols/icmpv4/tests.rs:35]: Could you file a bug on this issue (all tests that use unwrap or similar)?  Thanks.

File: src/protocols/ipv4/datagram.rs (hunk 2 above)
Comment (BrianZill) [reply src/protocols/ipv4/datagram.rs]: Okay, I think we resolved in our chat that you're going to restore this to the original parse function (or equivalent).

File: src/protocols/icmpv4/tests.rs (hunk 10 above)
Comment (ppenna) [reply src/protocols/icmpv4/tests.rs:35]: Issue created: https://github.com/demikernel/catnip/issues/119

File: src/protocols/ipv4/datagram.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/ipv4/datagram.rs]: Rolled back to `parse()` with `trim` and `adjust`

**Review by BrianZill — APPROVED** [review]
